test/sqllogictest/tpch.slt

```diff
@@ -1491,39 +1491,41 @@ Explained Query:
       Project (#1, #2) // { arity: 2 }
         Join on=(#0 = #3) type=differential // { arity: 4 }
           implementation
-            %0:l0[#0] » %1[#0]UKA
+            %1[#0]UKA » %0:l0[#0]KA
           ArrangeBy keys=[[#0]] // { arity: 3 }
             Get l0 // { arity: 3 }
           ArrangeBy keys=[[#0]] // { arity: 1 }
             Distinct group_by=[#0] // { arity: 1 }
               Project (#0) // { arity: 1 }
-                Filter (integer_to_numeric(#2) > (0.5 * #5)) // { arity: 6 }
+                Filter (integer_to_numeric(#2) > #5) // { arity: 6 }
                   Join on=(#0 = #4 AND #1 = #3) type=differential // { arity: 6 }
                     implementation
-                      %0:l1[#1, #0] » %1[#0, #1]UKKAf
-                    ArrangeBy keys=[[#1, #0]] // { arity: 3 }
+                      %1[#1, #0]UKKA » %0:l1[#0, #1]KKAf
+                    ArrangeBy keys=[[#0, #1]] // { arity: 3 }
                       Project (#0, #1, #3) // { arity: 3 }
                         Filter (#0 = #2) // { arity: 4 }
                           Get l1 // { arity: 4 }
-                    ArrangeBy keys=[[#0, #1]] // { arity: 3 }
-                      Reduce group_by=[#0, #1] aggregates=[sum(#2)] // { arity: 3 }
-                        Project (#0, #1, #6) // { arity: 3 }
-                          Filter (#12 >= 1995-01-01) AND (date_to_timestamp(#12) < 1996-01-01 00:00:00) // { arity: 18 }
-                            Join on=(#0 = #3 AND #1 = #4) type=differential // { arity: 18 }
-                              implementation
-                                %1:lineitem[#1, #2] » %0[#0, #1]UKKAiif
-                              ArrangeBy keys=[[#0, #1]] // { arity: 2 }
-                                Distinct group_by=[#0, #1] // { arity: 2 }
-                                  Project (#1, #2) // { arity: 2 }
-                                    Get l1 // { arity: 4 }
-                              ArrangeBy keys=[[#1, #2]] // { arity: 16 }
-                                Get materialize.public.lineitem // { arity: 16 }
+                    ArrangeBy keys=[[#1, #0]] // { arity: 3 }
+                      Project (#0, #1, #3) // { arity: 3 }
+                        Map ((0.5 * #2)) // { arity: 4 }
```

Comment:
This looks like a regression that we might want to grok. Previously we could use the output arrangement of `Reduce`, and now we add the `* 2` output column and create a new arrangement of the result.

Reply:
In the old plan, uniqueness is on the second input. In the new plan, uniqueness is on the first input, which we deem better. (Again, recall that the `JoinInputCharacteristics` of the first input is about the left side of the first join stage. This is different from other inputs, where the `JoinInputCharacteristics` is about the right side of a join stage.) This consideration trumps arrangement re-use (because `unique_key` is earlier than `arranged` in `JoinInputCharacteristics`, so a unique input being earlier is deemed better).

However, this doesn't make much sense, because uniqueness on the first input is exactly as good as uniqueness on the second input, so arrangement re-use should be the deciding factor instead. More generally, uniqueness on the left side of a join stage is exactly as good as on the right side, but currently we take it into account only on the right side. I opened an issue to tweak the meaning of `JoinInputCharacteristics::unique_key`: https://github.com/MaterializeInc/materialize/issues/17684

Reply:
Btw. we could also make `JoinImplementation` mindful about situations where an existing arrangement cannot be re-used only because of a different column order. In many cases, nothing would prevent us from reordering the keys of intermediate join stages to match an existing arrangement. For example, in the particular situation in this slt test, we could re-use the arrangement if we insert a reversing after `self.bound[rel].sort();`, because then the order happens to match the existing arrangement.